src/messaging/ExchangeContext.cpp

```diff
@@ -296,9 +296,6 @@ bool ExchangeContext::MatchExchange(SecureSessionHandle session, const PacketHea
         // The exchange identifier of the message matches the exchange identifier of the context.
         (mExchangeId == payloadHeader.GetExchangeID())
 
-        // AND The message was received from the peer node associated with the exchange
-        && (mSecureSession == session)
```

Comment:
This leaves a serious secure hold that a malicious attacker can inject a unsecure packet to a secure session.

Reply:
As we discussed offline, the exchange is not associated with session per spec, and the keyID within session here is not correct (hard coded to 0 to pass the exchange match).

We might need to come up with another solution out of spec to address this issue. Based on spec, when matching an exchange, we should only match source {fabricId, nodeId} with the exchange's desired {fabricId, nodeId} beside exchangeID.

Reply:
The message arrived in exchange layer has already been decoded, if the hacker inject a secure packet with wrong keyID or insecure message to a secure session, it will fail in decryption logic if the KeyID is wrong or the packet is not encrypted using the correct/corresponding key.

Reply:
> the exchange is not associated with session per spec

How can the above be true?  A message incoming over CASE must be matched to the session ID (local key ID) before forwarding to other layers, or the security model evaporates.  The code is not free to forward messages anywhere it pleases just  because the message got in the door.  Exchanges and application layer behavior do have specific associations with sessions and can only receive messages from their associated session.  This means the session ID needs to be associated with the exchange and checked, and if it doesn't match, the message should be discarded.

Without out this, you can maliciously inject messages into code paths that are intended for other peers.  This is a vector for permissions escalation.

Reply:
"An Exchange SHALL be bound to exactly one underlying session" was removed from the spec two month ago. What we need to decide is if an exchange is bonded to its underling session. Such as in CASE resumption case, do we need to drop the old exchange and create a new one, or just update the session.

https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/3588 makes it clear. The exchange is bonded to the underlying session by Session ID.

Reply:
If we decide to allow exchanges to persist beyond the lifetime of a session, we would still need to verify that the exchange is associated with the _current_ session.  This is the only way to carry our peer authentication model through the exchange layer.

Reply:
And come to think of it, I don't think that's really sufficient.  It does need to be the same session.  In my view, there is an abstract owner / peer in the exchange and this is identified by the session.  It's not just sufficient to authenticate the node.  What if the node is supporting multiple concurrent sessions with a peer?  This becomes nonsensical if an exchange can move to a different session ID.

Reply:
Clarified with Martin, there is only one secure session at a time per peer node within the the same fabric, all exchanges shared the same underling session, and when the session ends, we also need to tear down the associated exchange.
The idea is to make the policy simple in the spec to avoid various confusion in implementation.

Reply:
This all makes sense except for:

```
there is only one secure session at a time per peer node within the the same fabric
```

Why?  That's a very significant constraint.

Reply:
Per Martin, the goal is to simply processing of exchanges and bind them to exactly one session and transport. The only use case for muti-session is mutil-fabric, different credentials even to the same peer. For any fabric, the hierarchy of exchanges associating with one underlying session holds, They would associate with the sessions that matches the op cred being used for the interaction as per the ACL.

Do we have any scenario we need to use more than one secure sessions to the same peer node under the same fabric?

Reply:
What assumption is leading to the belief that associating an exchange to a single session means there can only be a single session at any given time between two given peers on a given fabric?  This constraint isn't necessary and it is very restrictive.

> Per Martin, the goal is to simply processing of exchanges and bind them to exactly one session and transport.

Great, that's simple and makes perfect sense.

> The only use case for muti-session is mutil-fabric, different credentials even to the same peer.

I disagree.  Why should the case of a single node opening two sessions to a given peer be any different than two separate nodes on the same fabric opening sessions to the given peer?  Nothing in the spec makes such a distinction necessary.  Each node manages its session ID space so it can properly associate and authenticate incoming messages with the correct internal context.

> For any fabric, the hierarchy of exchanges associating with one underlying session holds

Yes, I agree.  Now why can't I then have 10 concurrent exchanges mapping back to 10 sessions between the two peers?  The only thing that should prevent this is lack of capability / memory in one, or the other, or both peers.

Is it because there is an assumption that exchanges also somehow have node IDs as primary key?  You see, that there is the problem.  The whole point of the session ID is that there is an abstraction for nodes to recover the correct context from incoming messages, whatever that might be, and be able to authenticate the association between the incoming message and an internal context.

> They would associate with the sessions that matches the op cred being used for the interaction as per the ACL.

Yes, the session ID maps back to an encryption key, which in turn maps back to OpCreds.  How does this preclude two given peers from opening multiple sessions with one another?

> Do we have any scenario we need to use more than one secure sessions to the same peer node under the same fabric?

Yes, many and they are unknowable.  Currently, the spec can support this and the sdk ought to too.

Reply:
The change in this PR is based on assumption that an exchange is only bind to exactly one underlying session on a given fabric.

Regarding your question, "if we could open more than one session to peer node on a given fabric", I think there is more about policy definition than technical restriction and this is not disabled by this PR. Currently, this is not supported by our SDK. I hope we could clarify this explicitly from the spec side to avoid confusion in implementation.

Reply:
I just do not believe there is any ambiguity in this regard within the spec.  The spec is very clear that the session ID is what apps should use to recover context for incoming traffic. This implicitly allows for the multiple concurrent sessions between peers that I'm arguing should be supported.

But the security model also depends upon this, and not doing so is a good way to introduce bugs and security issues.  https://github.com/project-chip/connectedhomeip/pull/8441#issuecomment-883364306 alludes to a potential security issue, and #8481 shows us examples of bugs that can occur when we try to use things other than the session ID to locate context.

I am aware that the `chip::Controller::DeviceController` object is not architected to support multiple concurrent sessions with a single peer because it maintains a single session context per peer and because it demuxes exchange traffic by keying on node ID.  This is not necessarily broken from a security perspective because: 1) the controller is currently single-fabric, and 2) the controller uses a single session ID allocator in such a way that there is a 1:1 mapping between session IDs and node IDs.

But I'm trying to be very forceful in my advocacy that we should not preclude this type of usage at these lower levels, for instance within the exchange manager.  If an implementer wants to roll their own controller object that supports multiple concurrent sessions, this ought to be possible.  Code that would preclude such a thing does not make sense and is probably buggy (again, see #8481).

Reply:
My point is this PR does not enable or disable multiple session support to a peer on a given fabric, it uses session ID to match the exchange for incoming packet, it seems it is aligned with what you described above, please lets me know if my understanding is wrong.

Reply:
Yes, I was specifically arguing against this statement from https://github.com/project-chip/connectedhomeip/pull/8441/files#r671514752:

> there is only one secure session at a time per peer node within the the same fabric

I agree that the code in this PR does not enforce such a thing.

Getting back to the code that is actually in this PR, in fact I think this line is actually incorrect.  The session ID is the local key ID.  Shouldn't this be the following?

```
        // AND The Session ID associated with the incoming message matches the Session ID associated with the exchange.
        && (mSecureSession.GetLocalKeyId() == session.GetLocalKeyId())
```

I guess I also don't know why we have a delta here at all.  I would have thought the `SecureSessionHandle` == operator overload would be what we should be using.  But that also seems to be incorrect:

```
    bool operator==(const SecureSessionHandle & that) const
    {
        return mPeerNodeId == that.mPeerNodeId && mPeerKeyId == that.mPeerKeyId && mAdmin == that.mAdmin;
    }
```

Again, this should be the local key, no?  In other words, we should be matching on the number that came from our local session ID allocator.  That's what we authenticate against.

Reply:
That needs to be further specified in spec, currently, we only have PeerKeyId defined in session handle. Will update the PR as long as it is clarified

Reply:
In my reading of the spec, this is clear.  Each message has a session ID field.  This is plain text and tells the recipient which of its keys was used to encrypt the message and, therefore, informs the recipient of which key it should use to decrypt the message.  If the recipient can use this key to decrypt the message and the MIC shows integrity, the recipient has authentication.

I believe in the sdk, this is referred to as 'local key ID'.  The `SecureSessionHandle` does need the peer key to encrypt outgoing traffic, but it appears it's missing the local key ID, which is ultimately the thing that should be used for session matching.

Another way to look at this: I set my session ID (local key ID) and all my peers set their session IDs (peer key IDs).  I can't authenticate based upon peer key ID because I don't set this and it's not guaranteed unique.